onecodex/models/helpers.py

```diff
@@ -136,25 +138,47 @@ def _download(
         if path and file_obj:
             raise OneCodexException("Please specify only one of: path, file_obj")
 
-        if path is None and file_obj is None:
-            path = os.path.join(os.getcwd(), _filename)
-
-        if path and os.path.exists(path):
-            raise OneCodexException("{} already exists! Will not overwrite.".format(path))
-
         try:
             method_to_call = getattr(self._resource, _resource_method)
+            download_link_info = method_to_call()
+
+            if path is None and file_obj is None:
+                if _filename is None:
+                    if "save_as_filename" not in download_link_info:
+                        raise OneCodexException(
+                            "Please specify `path`, `file_obj`, or `_filename`."
+                        )
+                    _filename = download_link_info["save_as_filename"]
+                path = os.path.join(os.getcwd(), _filename)
+
+            if path and os.path.exists(path):
+                raise OneCodexException("{} already exists! Will not overwrite.".format(path))
 
             if use_potion_session:
-                resp = self._resource._client.session.get(
-                    method_to_call()["download_uri"], stream=True
-                )
+                session = self._resource._client.session
             else:
-                resp = requests.get(method_to_call()["download_uri"], stream=True)
+                session = requests.Session()
+
+            link = download_link_info["download_uri"]
+
+            # Retry up to 5 times with backoff timing of 2s, 4s, 8s, 16s, and 32s (applies to all
+            # HTTP methods). 404 is included for cases where the file is being asynchronously
+            # uploaded to S3 and is expected to be available soon.
+            retry_strategy = Retry(
+                total=5,
+                backoff_factor=2,
+                status_forcelist=[404, 429, 500, 502, 503, 504],
+                method_whitelist=False,
+            )
+            adapter = HTTPAdapter(max_retries=retry_strategy)
+            session.mount("http://", adapter)
+            session.mount("https://", adapter)
+
+            resp = session.get(link, stream=True)
 
             with (open(path, "wb") if path else file_obj) as f_out:
                 if progressbar:
-                    with click.progressbar(length=self.size, label=self.filename) as bar:
+                    with click.progressbar(length=self.size, label=self.id) as bar:
```

Comment:
what's the reason for that change?

Reply:
`self.filename` is the original filename (e.g. `assembly.fa.gz`) and might be confusing to users since they're getting an uncompressed FASTA (e.g. `Escherichia_coli_ABC_8.fasta`). The UUID isn't super informative to the user but at least it won't be misleading regarding the filename or file type. Open to other options here -- I wanted to use the actual filename we're downloading the file to, but the user may supply a file-like object, which isn't guaranteed to have a path/filename.

Reply:
Can we do:

1. see if there is the actual filename and use that
2. if not use `self.id` with a note in parenthesis explaining what is is

Reply:
I ended up using the actual filename we're downloading to if we have a path, otherwise using `self.filename` since that's probably more descriptive for users (and they are already managing their own file paths).